Add Fourier transform playground example

brisvag opened the pull request:
# Description

Adds a new example to play around with 2D sine waves and fourier transforms. I made it to better visualise and explain how power spectrum and phase behave depending on the fourier components.

The example is also a nice showcase of two-way threading communication :P

Here's what it looks like:

https://github.com/napari/napari/assets/23482191/522f3e8d-cbe5-41c4-9c78-9da05d751adc

brisvag committed:
add fourer transform playground

brisvag committed:
better order

brisvag committed:
fix yield and add napari.run

brisvag committed:
auto run first time and reset view on add/remvoe layer

brisvag committed:
better order and interp

stefanv commented:
Liking this very much!!

FFT visualization often benefits from taking the log, to avoid large changes in amplitude. Something like `np.log10(c + spectrum)` and then playing with the value of `c`, given the application.

jni commented on examples/fourier_transform_playground.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def wave_2d(wavelength, angle, phase_shift, speed):
+    """
+    Generate a 2D sine wave based on angle and wavelength.
```
```suggestion
    """Generate a 2D sine wave based on angle and wavelength.
```

jni commented on examples/fourier_transform_playground.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def update_layer(name, data, **kwargs):
+    """
+    Update a layer in the viewer with new data.
```
```suggestion
    """Update a layer in the viewer with new data.
```

jni commented on examples/fourier_transform_playground.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def combine_and_set_data(waves):
+    """
+    Merge 2D waves, calculate the FT and update the viewer.
+    """
```
```suggestion
    """Merge 2D waves, calculate the FT and update the viewer."""
```

jni commented on examples/fourier_transform_playground.py:
```diff
@@ -0,0 +1,150 @@
+    speed: float = 1,
+    run=True,
+):
+    """
+    Send new parameters to the listening thread to update the 2D waves.
```
```suggestion
    """Send new parameters to the listening thread to update the 2D waves.
```

jni reviewed (approved):
I have to say I was *very* confused about where the animation was coming from... Then I saw the sneaky `time()` call way inside `wave_2d`! That'll learn me to read docstrings 😂 *Personally*, I would prefer for `wave_2d` to be purely functional, take in the current time, and the higher-level loop (`update_viewer`) to pass in the time. But I wouldn't hold up the PR for it.

At any rate, I played with it, and it's so mesmerising.... I love it. 😍 😵‍💫 😃

brisvag committed:
Merge branch 'main' of github.com:napari/napari into example/ft-simulator

brisvag committed:
use anlge for phase

brisvag committed:
Merge branch 'example/ft-simulator' of github.com:brisvag/napari into example/ft-simulator

brisvag committed:
multiply phase by ps for better vis

brisvag committed:
frequency is more intuitive

brisvag committed:
move time out of wave func, use log for visualisation and fix phase

brisvag commented:
> FFT visualization often benefits from taking the log, to avoid large changes in amplitude. Something like np.log10(c + abs(spectrum)) and then playing with the value of c, given the application.

Nice! I settled on `c = 10` for this application, it improves the readability of the phase.

Speaking of which, I fixed the phase, since I was doing it wrong before: I was taking only the imaginary part instead of getting the angle of the complex number.

I also moved out the time component from the base function, which I agree makes things cleaner. Should be good to go!

GenevieveBuckley commented:
This is great!

It reminds me very strongly of Grant Jensen's [image Fourier transform talk](https://www.youtube.com/watch?v=nyk75ufbP74&list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5&index=17), which is part of his [incredible Cryo-EM lecture series](https://www.youtube.com/playlist?list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5) (seriously you should go look, it's brilliant and there are several other fourier transform videos in that series as well as the one I linked above)

Can we add links to those in the text description as well? I think that would really help boost people's understanding.

jni commented:
The docs build is failing, I presume because some of the threading / interactive examples need to be excluded from running in the gallery. (But tbh I don't know, the fail log doesn't offer much of a clue...) I can't remember how this works... @melissawm any ideas? 😬 🙏

brisvag commented:
> It reminds me very strongly of Grant Jensen's [image Fourier transform talk](https://www.youtube.com/watch?v=nyk75ufbP74&list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5&index=17), which is part of his [incredible Cryo-EM lecture series](https://www.youtube.com/playlist?list=PLhiuGaXlZZenm7lu5qv_A59zEWkRKkBn5)

Yeah, that's part of the go-to lecture series we always suggest to newbies (me included :P) here :) Being able to interact with it is so much better though. Explaining aliasing by moving a slider and seing the peaks in the power spectrum "bounce" on the edges is so satisfying :)

melissawm commented:
I tested locally and couldn't find any issues (docs build just fine), but I restarted the CI job yesterday and was watching - it seems like the examples grow progressively slower after this new one, maybe something is leaking? Not sure how to debug locally as I don't see any issues but will try to poke around. Sorry I'm not too familiar with the threading stuff!

brisvag committed:
quit thread

brisvag commented:
Adding `thread.quit()` seems to solve the docs bulding issue, but now there's some odd segfault in pyside6, which also seem to refer to threads... may be related?

jni commented on examples/fourier_transform_playground.py:
```diff
@@ -0,0 +1,151 @@
+viewer.window.add_dock_widget(wdg, area='bottom')
+wdg()
+
+napari.run()
```
I think this should be protected by `if __name__ == '__main__':`, too? I'm a bit confused why it even works LOL

brisvag replied:
The issue is that to define some of the callbacks, the thread needs to already be spawned... so I would have to stick a lot of stuff inside the `if __name__` block... not sure that helps much :P

jni commented on examples/fourier_transform_playground.py:
```diff
@@ -0,0 +1,151 @@
+
+napari.run()
+
+thread.quit()
```
nice catch 😅
